File: 0957-minimum-add-to-make-parentheses-valid/0957-minimum-add-to-make-parentheses-valid.cpp
class Solution {
public:
    int minAddToMakeValid(string s) {
        int extra=0,open=0;
        for(auto ch:s){
            if(ch=='('){
                open++;
            }
            else{
                if(open>0){
                    open--;
                }else{
                    extra++;
                }
            }
        }
        return extra+open;
    }
};
/*stack approach
class Solution {
public:
    int minAddToMakeValid(string s) {
        stack<char> st;
        int count;
        for(auto ch:s){
            if(ch==')'){
                if(!st.empty() && st.top()=='('){
                    st.pop();
                    continue;
                }
            }
            st.push(ch);
        }
        return st.size();
    }
};*/

File: 0957-minimum-add-to-make-parentheses-valid/0957-minimum-add-to-make-parentheses-valid.py
class Solution(object):
    def minAddToMakeValid(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack = []

        for char in s:
            if char == '(':
                stack.append(char)
            elif char == ')':
                if stack and stack[-1] == '(':
                    stack.pop()  # matched
                else:
                    stack.append(char)  # unmatched ')'

        return len(stack)